### PR TITLE
Insert ESM exports for CommonJS entries

### LINKED ISSUE #3586: Dynamically importing CommonJS produces empty bundle
Dynamically importing a commonjs module from a `esmodule` asset produces an empty bundle.

```html
<script type="module" src="./index.js"></script>
```
```js
// index.js
import("./commonjs.js").then(v => console.log(v));
```
```js
// commonjs.js
module.exports = 2;
```

Output (why is that even called `...exports`?):
```js
var $c0ee4c22c4f80f14efb8476eb9807b50$exports = {};
$c0ee4c22c4f80f14efb8476eb9807b50$exports = import('' + './commonjs.e4b0f555.js');
$c0ee4c22c4f80f14efb8476eb9807b50$exports.then(v => console.log(v));
```

`./commonjs.e4b0f555.js` is empty, without `treeShake`, the output is
```js
// ...
var $dfc1ea661454a5b57fd7e754dd151b32$exports = {};
$dfc1ea661454a5b57fd7e754dd151b32$exports = 2;
```

We check `b.env.outputFormat === 'esmodule'`, but this value is just inherited from that parent asset.

The root cause is of course that we don't/can't translate commonjs to esmodules, but this is a case where this might be unexpected.

cc @devongovett

### PR DESCRIPTION
**Pull request by mischnic**
# ↪️ Pull Request

Closes #3586 :

> ```html
> <script type="module" src="./index.js"></script>
> ```
> ```js
> // index.js
> import("./commonjs.js").then(v => console.log(v));
> ```
> ```js
> // commonjs.js
> module.exports = 2;
> ```
> 
> Output:
> `./commonjs.e4b0f555.js` is empty
> without `treeShake`, the output is
> ```js
> // ...
> var $dfc1ea661454a5b57fd7e754dd151b32$exports = {};
> $dfc1ea661454a5b57fd7e754dd151b32$exports = 2;
> ```

This essentially exports `module.exports` as the `default` export in commonjs bundles with `outputFormat: "esmodule"`

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>packages/benchmarks/kitchen-sink</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 10.34s | -313.00ms |
| Cached | 8.09s | -317.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.ff215454.webp | 102.94kb | +0.00b | 227.00ms | +25.00ms ⚠️ |
| dist/modern/parcel.76ee4591.webp | 102.94kb | +0.00b | 226.00ms | +25.00ms ⚠️ |
| dist/modern/index.css | 29.00b | +0.00b | 1.18s | -153.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.ff215454.webp | 102.94kb | +0.00b | 14.00ms | +1.00ms ⚠️ |
| dist/modern/parcel.76ee4591.webp | 102.94kb | +0.00b | 13.00ms | +1.00ms ⚠️ |

</p></details>

<details><summary>packages/benchmarks/react-hn</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 15.19s | -174.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/245871934644945417)**

File: packages/shared/scope-hoisting/src/formats/esmodule.js [hunk 1]
```diff
@@ -92,6 +96,10 @@ export function generateExports(
 
       exportedIdentifiers.set(symbol, exportSymbol);
     }
+
+    if (entry.meta.hasCommonJSExports && !entry.meta.isES6Module) {
+      applySyntheticCommonJSExport = true;
```
Comment (mischnic) [inline packages/shared/scope-hoisting/src/formats/esmodule.js]: Is this logic correct?

File: packages/core/integration-tests/test/integration/formats/esm-async-commonjs/index.js [hunk 2]
```diff
@@ -0,0 +1 @@
+output = import('./async').then(a => a.foo + 2);
```
Comment (devongovett) [inline packages/core/integration-tests/test/integration/formats/esm-async-commonjs/index.js]: How would you access `module.exports` here? `a` should be the string `"async"` right? Looks like with the logic below you'd actually have to do `a.default` though...

Comment (devongovett) [reply packages/core/integration-tests/test/integration/formats/esm-async-commonjs/index.js]: I guess that's not really possible with ESM though...

**Comment by mischnic** [comment]
Superseded by: #3800

**Pull request closed without merging**